File: backend/src/main/java/com/primatoos/backend/model/WorkOrder.java
package com.primatoos.backend.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "work_orders")
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class WorkOrder {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "order_number", nullable = false, unique = true, updatable = false)
    private Long orderNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @Column(nullable = false)
    private LocalDate date;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "responsible_user_id", nullable = false)
    private User responsibleUser;

    @Column(nullable = false)
    private String stage;

    private String location;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(name = "daily_goal", columnDefinition = "TEXT")
    private String dailyGoal;

    @Column(name = "planned_start_time")
    private LocalTime plannedStartTime;

    @Column(name = "planned_end_time")
    private LocalTime plannedEndTime;

    @Column(name = "materials_needed", columnDefinition = "TEXT")
    private String materialsNeeded;

    @Column(columnDefinition = "TEXT")
    private String tools;

    @Column(name = "safety_guidelines", columnDefinition = "TEXT")
    private String safetyGuidelines;

    @Column(name = "quality_criteria", columnDefinition = "TEXT")
    private String qualityCriteria;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private WorkOrderStatus status;

    @Builder.Default
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "work_order_workers",
            joinColumns = @JoinColumn(name = "work_order_id"),
            inverseJoinColumns = @JoinColumn(name = "worker_id")
    )
    private Set<Worker> assignedWorkers = new HashSet<>();

    @Setter(AccessLevel.NONE)
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }
}

File: frontend/src/types/workOrder.ts
import type { UserSummaryResponse } from './common'
import type { ProjectSummaryResponse } from './project'
import type { WorkerSummaryResponse } from './worker'

export type WorkOrderStatus = 'DRAFT' | 'RELEASED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED'

export interface WorkOrderResponse {
  id: number
  orderNumber: number
  project: ProjectSummaryResponse
  date: string
  responsibleUser: UserSummaryResponse
  stage: string
  location: string | null
  description: string
  dailyGoal: string | null
  plannedStartTime: string | null
  plannedEndTime: string | null
  materialsNeeded: string | null
  tools: string | null
  safetyGuidelines: string | null
  qualityCriteria: string | null
  notes: string | null
  status: WorkOrderStatus
  assignedWorkers: WorkerSummaryResponse[]
  createdAt: string
}

export interface WorkOrderSummaryResponse {
  id: number
  orderNumber: number
  stage: string
}

export interface WorkOrderCreateRequest {
  projectId: number
  date: string
  responsibleUserId: number
  stage: string
  location?: string | null
  description: string
  dailyGoal?: string | null
  plannedStartTime?: string | null
  plannedEndTime?: string | null
  materialsNeeded?: string | null
  tools?: string | null
  safetyGuidelines?: string | null
  qualityCriteria?: string | null
  notes?: string | null
  assignedWorkerIds?: number[]
}

export interface WorkOrderUpdateRequest {
  projectId: number
  date: string
  responsibleUserId: number
  stage: string
  location?: string | null
  description: string
  dailyGoal?: string | null
  plannedStartTime?: string | null
  plannedEndTime?: string | null
  materialsNeeded?: string | null
  tools?: string | null
  safetyGuidelines?: string | null
  qualityCriteria?: string | null
  notes?: string | null
  assignedWorkerIds?: number[]
}
